core/src/main/scala/cr/pulsar/Consumer.scala

```diff
@@ -71,40 +69,40 @@ object Consumer {
   case class Message[A](id: MessageId, key: MessageKey, payload: A)
   case class DecodingFailure(msg: String) extends Exception(msg) with NoStackTrace
 
-  private def mkConsumer[F[_]: Async: Spawn, E: Schema](
+  private def mkConsumer[F[_]: Sync: FutureLift, E: Schema](
       client: Pulsar.T,
       sub: Subscription,
       topic: Topic,
       opts: Options[F, E]
   ): Resource[F, Consumer[F, E]] = {
-    val acquire =
-      F.delay {
-        val c = client.newConsumer(E.schema)
-        val z = topic match {
-          case s: Topic.Single => c.topic(s.url.value)
-          case m: Topic.Multi  => c.topicsPattern(m.url.value.r.pattern)
-        }
-        opts.deadLetterPolicy
-          .fold(z)(z.deadLetterPolicy)
-          .readCompacted(opts.readCompacted)
-          .subscriptionType(sub.`type`.pulsarSubscriptionType)
-          .subscriptionName(sub.name.value)
-          .subscriptionMode(sub.mode.pulsarSubscriptionMode)
-          .subscriptionInitialPosition(opts.initial)
-          .subscribeAsync
-      }.futureLift
+
+    val acquire = F.futureLift {
+      val c = client.newConsumer(E.schema)
```

Comment:
Creating a new consumer is an effectful operation. Either we suspend this or we extract this out as a new capability trait such as `MkConsumer[F].newConsumer(E.schema)`. Did something similar here: https://github.com/gvolpe/pfps-shopping-cart/blob/second-edition/modules/core/src/main/scala/shop/resources/MkHttpClient.scala#L10

Reply:
Is it actually effectful? `newConsumer` returns a builder, which then is being built below by `subscribeAsync` which is wrapped into `futureLift`

Reply:
yeah you're right but `subscribeAsync` is definitely effectful and that's not being suspended, the future is being converted straight away

```scala
def futureLift[A](fa: CompletableFuture[A]): F[A]
```

Reply:
In the latest version I've made `futureLift` also suspend effects. Maybe `FutureLift` isn't the right name for it, but that's exactly what we are doing all the time.

Reply:
Ah I see, there's a tweak needed then, the parameter needs to be by-name, otherwise Scala will evaluate it before it is suspended...

```scala
def futureLift[A](fa: => CompletableFuture[A]): F[A]
```

Reply:
Indeed! Totally missed that, thanks